File: 79-word-search/word-search.cpp
class Solution {
public:
    vector<int> dir= vector<int>({-1,0,1,0,-1});
    bool dfs(vector<vector<char>>& board, int r, int c, int m, int n, string &word, int idx, vector<vector<bool>> &vis){
        if(idx== word.size()) return true;
        if(r<0 || c<0 || r==m || c==n || word[idx]!=board[r][c] || vis[r][c]) return false;
        vis[r][c]=true;
        for(int i=0;i<4;++i){
            int nr=r+dir[i], nc=c+dir[i+1];
            if(dfs(board, nr, nc, m, n, word, idx+1, vis)) return true;
        }
        vis[r][c]=false;
        return false;
    }
    bool exist(vector<vector<char>>& board, string word) {
        bool ans=false;
        int m= (int)board.size(), n= (int)board[0].size();
        vector<vector<bool>> vis(m, vector<bool> (n));
        for(int i=0;i<m;++i){
            for(int j=0;j<n;++j){
                if(board[i][j]==word[0]){
                    if(dfs(board, i, j, m, n, word, 0, vis)) return true;
                }
            }
        }
        return false;
    }
};

File: 79-word-search/word-search.java
class Solution {
    int[] dir= {-1,0,1,0,-1};
    private boolean check(char[][] board, int r, int c, String word, int pos, boolean used[][]){
        if(r<0 || c<0 || r==board.length || c==board[0].length || used[r][c]) return false;
        if(board[r][c]==word.charAt(pos)){
            if(pos==word.length()-1) return true;
            used[r][c]= true;
            boolean ans= false;
            for(int i=0;i<4;i++){
                if(check(board, r+dir[i], c+dir[i+1], word, pos+1, used)){
                    ans= true;
                    break;
                }
            }
            used[r][c]= false;
            return ans;
        }
        else return false;
    }
    public boolean exist(char[][] board, String word) {
        int n= board.length, m= board[0].length;
        boolean used[][]= new boolean[n][m];
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(board[i][j]==word.charAt(0) && check(board, i, j, word, 0, used)){
                    return true;
                }
            }
        }
        return false;
    }
}